xarray/core/dtypes.py

```diff
@@ -1,188 +1 @@
-        fill_value = np.nan
-
-    dtype = np.dtype(dtype)
-    fill_value = dtype.type(fill_value)
-    return dtype, fill_value
-
-
-NAT_TYPES = {np.datetime64("NaT").dtype, np.timedelta64("NaT").dtype}
-
-
-def get_fill_value(dtype):
-    """Return an appropriate fill value for this dtype.
-
-    Parameters
-    ----------
-    dtype : np.dtype
-
-    Returns
-    -------
-    fill_value : Missing value corresponding to this dtype.
-    """
-    _, fill_value = maybe_promote(dtype)
-    return fill_value
-
-
-def get_pos_infinity(dtype, max_for_int=False):
-    """Return an appropriate positive infinity for this dtype.
-
-    Parameters
-    ----------
-    dtype : np.dtype
-    max_for_int : bool
-        Return np.iinfo(dtype).max instead of np.inf
-
-    Returns
-    -------
-    fill_value : positive infinity value corresponding to this dtype.
-    """
-    if issubclass(dtype.type, np.floating):
-        return np.inf
-
-    if issubclass(dtype.type, np.integer):
-        if max_for_int:
-            return np.iinfo(dtype).max
-        else:
-            return np.inf
-
-    if issubclass(dtype.type, np.complexfloating):
-        return np.inf + 1j * np.inf
-
-    return INF
-
-
-def get_neg_infinity(dtype, min_for_int=False):
-    """Return an appropriate positive infinity for this dtype.
-
-    Parameters
-    ----------
-    dtype : np.dtype
-    min_for_int : bool
-        Return np.iinfo(dtype).min instead of -np.inf
-
-    Returns
-    -------
-    fill_value : positive infinity value corresponding to this dtype.
-    """
-    if issubclass(dtype.type, np.floating):
-        return -np.inf
-
-    if issubclass(dtype.type, np.integer):
-        if min_for_int:
-            return np.iinfo(dtype).min
-        else:
-            return -np.inf
-
-    if issubclass(dtype.type, np.complexfloating):
-        return -np.inf - 1j * np.inf
-
-    return NINF
-
-
-def is_datetime_like(dtype):
-    """Check if a dtype is a subclass of the numpy datetime types"""
-    return np.issubdtype(dtype, np.datetime64) or np.issubdtype(dtype, np.timedelta64)
-
-
-def result_type(
-    *arrays_and_dtypes: np.typing.ArrayLike | np.typing.DTypeLike,
-) -> np.dtype:
-    """Like np.result_type, but with type promotion rules matching pandas.
-
-    Examples of changed behavior:
-    number + string -> object (not string)
-    bytes + unicode -> object (not unicode)
-
-    Parameters
-    ----------
-    *arrays_and_dtypes : list of arrays and dtypes
-        The dtype is extracted from both numpy and dask arrays.
-
-    Returns
-    -------
-    numpy.dtype for the result.
-    """
-    types = {np.result_type(t).type for t in arrays_and_dtypes}
-
-    for left, right in PROMOTE_TO_OBJECT:
-        if any(issubclass(t, left) for t in types) and any(
-            issubclass(t, right) for t in types
-        ):
-            return np.dtype(object)
-
-    return np.result_type(*arrays_and_dtypes)
+from xarray.namedarray.dtypes import *  # noqa: F401, F403
```

Comment:
to minimize the impact of this reorganization and maintain backward compatibility with existing codebases, i am choosing to keep this module and import everything. 

i am curious to know if there are any alternative approaches to this situation or if we should remove this module altogether and address the issue head-on.

Reply:
> n and maintain backward compatibility with existing codebases

This is all private API so I think we should just delete.

Reply:
I was thinking copy/paste dtypes.py and in NamedArray only save what's necessary. Then you wont have to deal with the downstream headaches.
Once the NamedArray design has settled; clean up the xarray dtypes-version.

`namedarray.dtypes` probably shouldn't be public to xarray either.

Reply:
thank you for the clarification, @Illviljan! copy/paste plus latter cleanup of xarray dtypes-version sounds good.